### PR TITLE
Create register mechanism for op implementation + runtime check of of op definition vs op results

### PR DESCRIPTION
**Pull request by achoum**
Details:

- Op implementations are registered similarly as op definitions.
- Unit tests ensure that the expected op implementation / definition are available.
- Runtime check input and output of op implementation against the op definition.

For later:

- Fix: The assign operator always output float64 values (instead of returning values of the same type as the input events).

Note: This PR is starting from the "gbm_normalize_tests" branch / [PR#42](https://github.com/google/temporian/pull/42).

### CODE REVIEW

File: temporian/core/evaluator.py [hunk 1]
```diff
@@ -33,15 +33,23 @@
 def evaluate(
     query: Query,
     input_data: Data,
-    backend: AvailableBackends = "numpy",
+    verbose: int = 0,
+    check_execution: bool = True,
```
Comment (ianspektor) [inline temporian/core/evaluator.py:37]: Default to False / add TODO to change it to False before public release?

File: temporian/implementation/numpy/test/registered_operators_test.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+class RegisteredOperatorsTest(absltest.TestCase):
+    def test_base(self):
+        expected_implementations = [
+            "ARITHMETIC",
+            "SIMPLE_MOVING_AVERAGE",
+            "CALENDAR_DAY",
+            "ASSIGN",
+            "SELECT",
+            "PROPAGATE",
+            "LAG",
```
Comment (ianspektor) [inline temporian/implementation/numpy/test/registered_operators_test.py:31]: same nit about alphabetical order

**Review by ianspektor — APPROVED** [review]

File: temporian/implementation/numpy/operators/assign.py [hunk 3]
```diff
@@ -87,6 +88,8 @@ def __call__(
                 for feature in right_event.data[index]:
                     new_feature = NumpyFeature(
                         name=feature.name,
+                        # TODO: Bug: The dtype is set to float64 independently
+                        # of the dtype in the op definition.
```
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/implementation/numpy/operators/assign.py:92]: https://github.com/google/temporian/issues/47

File: temporian/core/data/feature.py [hunk 4]
```diff
@@ -28,6 +28,11 @@ def __init__(
         sampling: Optional[sampling_lib.Sampling] = None,
         creator: Optional[Any] = None,
     ):
+        assert isinstance(name, str), f"Got {name}"
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/data/feature.py:31]: I think this is the limit of programming in python. The language does not enforce typing, and it seems strange to check programmatically the type of all the arguments.

Maybe we could think about a way to do it clearly and consistently. For example, by creating (or reusing, if this exist already) a utility for type checking, and to apply it to all the API surface.

*Adding todo

File: temporian/core/data/feature.py [hunk 5]
```diff
@@ -28,6 +28,11 @@ def __init__(
...
+        assert sampling is None or isinstance(
+            sampling, sampling_lib.Sampling
+        ), f"Got {sampling}"
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/data/feature.py:34]: *Covered by the todo above.
*Adding item to the agenda of the next meeting.

File: temporian/core/data/sampling.py [hunk 6]
```diff
@@ -23,6 +23,8 @@ def __init__(
         index: List[str],
         creator: Optional[Any] = None,
     ):
+        assert isinstance(index, list), f"Got {index}"
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/data/sampling.py:26]: 🐟 ... lol

File: temporian/core/evaluator.py [hunk 7]
```diff
@@ -33,15 +33,23 @@
...
 ) -> Dict[Event, Any]:
     """Evaluates a query on data.
 
     Args:
         query: Events to compute. Support event, dict and list of events.
         input_data: Dictionary of event and event values to use for the
           computation.
-        backend: Type of backend to use.
+        verbose: If >0, prints details about the execution on the standard error
+          output. The larger the number, the more information is displayed.
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/evaluator.py:46]: Imagine a user creating a script that print the events on the stdout (possibly in realtime). We dont want warnings to come and pollute the stream of events. stderr seems like the right place for warnings.

*Adding to next agenda

File: temporian/__init__.py [hunk 8]
```diff
@@ -29,16 +29,25 @@
...
+# from temporian.core.operators.arithmetic import divide
+# from temporian.core.operators.arithmetic import multiply
+# from temporian.core.operators.arithmetic import substract
+# from temporian.core.operators.arithmetic import sum
+# from temporian.core.operators.assign import assign
+# from temporian.core.operators.calendar_day_of_month import calendar_day_of_month
+# from temporian.core.operators.lag import lag
+# from temporian.core.operators.lag import leak
+# from temporian.core.operators.select import select
+# from temporian.core.operators.simple_moving_average import sma
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/__init__.py:44]: *Removed

File: temporian/core/test/registered_operators_test.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+class RegisteredOperatorsTest(absltest.TestCase):
+    def test_base(self):
+        expected_operators = [
+            "ARITHMETIC",
+            "CALENDAR_DAY",
+            "ASSIGN",
+            "SELECT",
+            "PROPAGATE",
+            "LAG",
+            "SIMPLE_MOVING_AVERAGE",
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/test/registered_operators_test.py:32]: Good point.
*Sorted operator names alphabetically.

File: temporian/core/evaluator.py (hunk 1 above)
Comment (achoum) [reply temporian/core/evaluator.py:37]: Sure.

*Default to False + add todo.

File: temporian/core/data/feature.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/core/data/feature.py:33]: Maybe I wasn't expressive enough in my comment, sorry - I just meant changing the failed assertion message from `"Got {name}"` to something like `"name must be a string, but received {name}"`

File: temporian/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/__init__.py]: Created issue https://github.com/google/temporian/issues/56

File: temporian/core/data/feature.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/data/feature.py:33]: My bad.
*Improved error message.

**Comment by achoum** [comment]
Thanks!

**Pull request merged**